File: maze.cpp
//time : O(m*n)
//space: O(m*n)

class Solution {
public:
    
    bool hasPath(vector<vector<int>>& maze, vector<int>& start, vector<int>& destination) {
        int m = maze.size(), n = maze[0].size();
        vector<vector<int>> dirs = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};
        queue<vector<int>> queue;
        vector<vector<bool>> visa(m, vector<bool>(n, false));
        queue.push(start);
        visa[start[0]][start[1]] = true;

        while (!queue.empty()) {
            vector<int> curr = queue.front();
            queue.pop();
            for (const auto& dir : dirs) {
                int r = curr[0];
                int c = curr[1];
                while (r >= 0 && c >= 0 && r < m && c < n && maze[r][c] != 1) {
                    r += dir[0];
                    c += dir[1];
                }
                r -= dir[0];
                c -= dir[1];

                if (visa[r][c])
                    continue;

                if (r == destination[0] && c == destination[1]) {
                    return true;
                }

                visa[r][c] = true;
                queue.push({r, c});
            }
        }

        return false;
        
    }
};

File:  maze.java
//Time Complexity: O(mn)
//Spce Complexity: O(n)

public class maze {
    public static boolean hasPath(int[][] maze, int[]start, int[]end){
        if(start[0]==end[0] && start[1]==end[1]) return true;
        int m = maze.length;
        int n = maze[0].length;
        Queue<Integer> q = new LinkedList<>();
        int[][] dirs = {{0,1},{1,0},{-1,0},{0,-1}} ;
        q.add(start[0]);
        q.add(start[1]);
        maze[start[0]][start[1]] = 2;
        while(!q.isEmpty()){
            int r = q.poll();
            int c = q.poll();
            for(int[] dir: dirs){
                int i = r; int j = c;
                while(i>=0 && i<m && j>=0 && j<n && maze[i][j]!=1){
                    i+=dir[0];
                    j+=dir[1];
                }
                i-=dir[0];
                j-=dir[1];
                if(maze[i][j]!=2){
                    if(i==end[0] && j==end[1]) return true;
                    q.add(i); q.add(j);
                    maze[i][j] = 2;
                }
            }
        }
        return false;
    }
}